R/get_gene_expression.R

```diff
@@ -169,32 +180,68 @@ get_gene_expression = function(these_samples_metadata,
           dplyr::filter(ensembl_gene_id %in% ensembl_gene_ids) %>%
           as.data.frame() %>%
           pivot_wider(names_from = ensembl_gene_id, values_from = expression)
-
       }
-
     }
   }
 
   if(join_with == "mrna" & missing(expression_data)){
-    expression_wider = dplyr::select(wide_expression_data, -biopsy_id, -genome_sample_id)
-    expression_wider = left_join(these_samples_metadata, expression_wider, by = c("sample_id" = "mrna_sample_id"))
-
-    }else if(join_with == "genome" & missing(expression_data)){
-    expression_wider = dplyr::select(wide_expression_data, -mrna_sample_id, -biopsy_id) %>% dplyr::filter(genome_sample_id != "NA")
-    expression_wider = left_join(these_samples_metadata, expression_wider, by = c("sample_id" = "genome_sample_id"))
-
-    }else if(join_with == "any" & missing(expression_data)){
-    expression_wider = dplyr::select(wide_expression_data, -mrna_sample_id, -genome_sample_id)
-    expression_wider = left_join(these_samples_metadata, expression_wider, by = c("biopsy_id" = "biopsy_id"))
+    these_samples_metadata = dplyr::select(these_samples_metadata, sample_id)
+    expression_wider = dplyr::select(wide_expression_data, -biopsy_id) %>% 
+      left_join(these_samples_metadata, ., by = c("sample_id" = "mrna_sample_id"))
+    
+  }else if( (join_with == "genome" | join_with == "capture") & missing(expression_data) ){
+    
+    these_samples_metadata = dplyr::select(these_samples_metadata, sample_id, 
+                                           patient_id, biopsy_id)
+    expression_wider = left_join(these_samples_metadata, wide_expression_data,
+                                 by = c("patient_id", "biopsy_id"))
 
-    }else if(join_with == "mrna" & !missing(expression_data)){
-      expression_wider = wide_expression_data
-
-    }else if(join_with == "genome" & !missing(expression_data)){
-      expression_wider = wide_expression_data
-
-    }else if(join_with == "any" & !missing(expression_data)){
-      expression_wider = wide_expression_data
+    # add column to inform if there are more than one mrna sample id for a genome sample id
+    expression_wider = split(expression_wider$mrna_sample_id, expression_wider$sample_id) %>% 
+      .[lengths(.) > 1] %>% 
+      names %>% 
+      { mutate(expression_wider, multi_exp = ifelse(sample_id %in% ., 1, 0)) }
+    
+    ### filter out duplicated expressions based on prioritize_rows_by
+    if( !missing(prioritize_rows_by) & any(expression_wider$multi_exp == 1) ){
+      
+      prioritization_colNames = names(prioritize_rows_by)
```

Comment:
I think there needs to be a bit more flexibility here. For example PolyA data includes several different "protocol" fields: 
```
> count(expr_file, protocol)
  protocol                               n
1 SLX_Transcriptome                1019881
2 Strand_Specific_Transcriptome_2 25856983
3 Strand_Specific_Transcriptome_3  7019181
4 Transcriptome_Lite_2             3239622
```
It would be better to allow users to provide a list of vectors, e.g. 
`prioritize_rows_by = list(protocol = c("Strand_Specific_Transcriptome_2", "Strand_Specific_Transcriptome_3"))`

Then I think it needs to be set up to iterate over the names of the list items in `prioritize_rows_by`.

Reply:
What a good idea! Thanks!
It's done.